### PR TITLE
Chatkit SDK event driven

### PR DESCRIPTION
**Pull request by pablisco**
This PR will break on CI since the app is not updated yet and the platform dependencies are not deployed on `jcenter` either but would like to get some early feedback.

### CODE REVIEW

File: chatkit/src/main/java/com/pusher/chatkit/UserSubscription.kt [hunk 1]
```diff
@@ -81,98 +82,51 @@ class UserSubscription(
...
+            .map { (eventName, _, _, data) ->
+                when (eventName) {
+                    "initial_state" -> data.parseAs<InitialState>().map { handleInitialState(it) }
+                    "added_to_room" -> data.parseAs<RoomEvent>().map { handleAddedToRoom(it.room) }
+                    "removed_from_room" -> data.parseAs<RoomIdEvent>().map { handleRemovedFromRoom(it.roomId) }
+                    "room_updated" -> data.parseAs<RoomEvent>().map { handleRoomUpdated(it.room) }
+                    "room_deleted" -> data.parseAs<RoomIdEvent>().map { handleRoomDeleted(it.roomId) }
+                    "user_joined" -> data.parseAs<UserChangeEvent>().map { handleUserJoined(userId, it.roomId) }
+                    "user_left" -> data.parseAs<UserChangeEvent>().map { handleUserLeft(userId, it.roomId) }
+                    else -> throw Error("Invalid event name: $eventName")
```
Comment (luismfonseca) [inline chatkit/src/main/java/com/pusher/chatkit/UserSubscription.kt]: `Unknown event name`

Which begs the question: how will graceful will this SDK handle new events being added?

File: chatkit/src/main/java/com/pusher/chatkit/UserSubscription.kt [hunk 2]
```diff
@@ -81,98 +82,51 @@ class UserSubscription(
...
-        }
     }
 
     private fun handleUserLeft(userId: String, roomId: Int) {
-        userStore.findOrGetUser(
-            id = userId,
-            userListener = UserListener { user ->
+        userStore.findOrGetUser(id = userId)
+            .fold({ error ->
+                broadcastError(error, "User left a room but I failed getting it: $userId")
```
Comment (luismfonseca) [inline chatkit/src/main/java/com/pusher/chatkit/UserSubscription.kt]: the _I_ feels very personal...

File: chatkit/src/main/java/com/pusher/chatkit/messages/MessageService.kt [hunk 3]
```diff
@@ -1,16 +1,34 @@
...
+import com.pusher.platform.network.Promise
+import com.pusher.platform.network.asPromise
+import com.pusher.util.*
+import elements.Error
+import elements.Errors
 import elements.Subscription
 import kotlinx.coroutines.experimental.channels.ReceiveChannel
 
+typealias MessageListFutureResult = Promise<Result<List<Message>, Error>>
+typealias MessageIdResult = Promise<Result<Int, Error>>
```
Comment (luismfonseca) [inline chatkit/src/main/java/com/pusher/chatkit/messages/MessageService.kt]: `MessageIdFutureResult`?

File: chatkit/src/main/java/com/pusher/chatkit/messages/MessageService.kt [hunk 4]
```diff
@@ -45,23 +63,47 @@ class MessageService(
...
-                callback(MessageSentEvent.Failed(error))
-            }
-        )
+    fun sendMessage(
+        text: String? = null,
+        attachment: GenericAttachment = NoAttachment
+    ): MessageIdResult = when (attachment) {
+        is DataAttachment -> uploadFile(attachment, room.id)
+        is LinkAttachment -> Promise.now(AttachmentBody.Resource(attachment.link, attachment.type).asSuccess<AttachmentBody, elements.Error>())
+        is NoAttachment -> Promise.now(AttachmentBody.None.asSuccess<AttachmentBody, Error>())
```
Comment (luismfonseca) [inline chatkit/src/main/java/com/pusher/chatkit/messages/MessageService.kt]: Do you want to create a `Result.promiseSuccess<AttachementBody, Error>(AttachmentBody.None)`? :p

Comment (pablisco) [reply chatkit/src/main/java/com/pusher/chatkit/messages/MessageService.kt]: I'll put this as a task, as it touches the platform

**Pull request merged**